text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,527 @@
+
+We could not do this, and continue to add accessor functions to the `Error`
+trait whenever a new type reaches critical levels of popularity in error
+reporting.
+
+If we choose to do nothing we will continue to see hacks around the current
+limitations on the `Error` trait such as the `Fail` trait, which added the
+missing function access methods that didn't previously exist on the `Error`
+trait and type erasure / unnecessary boxing of errors to enable downcasting to
+extract members.
+[[1]](https://docs.rs/tracing-error/0.1.2/src/tracing_error/error.rs.html#269-274).
+
+
+## Use an alternative proposal that relies on the `Any` trait for downcasting
+
+```rust
+pub trait Error {
+    /// Provide an untyped reference to a member whose type matches the provided `TypeId`.
+    ///
+    /// Returns `None` by default, implementors are encouraged to override.
+    fn provide_context(&self, ty: TypeId) -> Option<&dyn Any> {
+        None
+    }
+}
+
+impl dyn Error {
+    /// Retrieve a reference to `T`-typed context from the error if it is available.
+    pub fn context<T: Any>(&self) -> Option<&T> {
+        self.provide_context(TypeId::of::<T>())?.downcast_ref::<T>()
+    }
+}
+```
+
+### Why isn't this the primary proposal?
+
+There are two big issues with using the `Any` trait that I believe justify the
+more complicated solution.
+
+- You cannot return dynamically sized types as `&dyn Any`
+- It's easy to introduce runtime errors with `&dyn Any` by either comparing to
+  or returning the wrong type
+
+By making all the type id comparison internal to the `Request` type it is
+impossible to compare the wrong type ids. By encouraging explicit type
+parameters when calling `provide` the compiler is able to catch errors where
+the type passed in doesn't match the type that was expected. So while the API
+for the main proposal is more complicated it should be less error prone.
+
+# Prior art
+[prior-art]: #prior-art
+
+I do not know of any other languages whose error handling has similar
+facilities for accessing members when reporting errors. For the most part,
+prior art for this proposal comes from within rust itself in the form of
+previous additions to the `Error` trait.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* What should the names of these functions be?
+    * `context`/`context_ref`/`provide_context`/`provide_context`
+    * `member`/`member_ref`
+    * `provide`/`request`
+* Should there be a by-value version for accessing temporaries?
+    * We bring this up specifically for the case where you want to use this
+      function to get an `Option<&[&dyn Error]>` out of an error, in this case,
+      it is unlikely that the error behind the trait object is actually storing
+      the errors as `dyn Error`s, and theres no easy way to allocate storage to
+      store the trait objects.
+* How should context handle failed downcasts?
+    * suggestion: panic, as providing a type that doesn't match the typeid
```

Comment:
What's the alternative? Map it to `None` as if no context was provided?

Reply:
The problem isn't when they return None, its when they return the wrong type

```rust
if TypeId::of::<SpanTrace>() == id {
   Some(&self.backtrace as &dyn Any)
}
```

This is pretty clearly a program error imo, this is only a problem with the `Any` based approach though, so I should probably remove this one. With `Request` your only options are to either do `request.provide::<SpanTrace>(&self.backtrace)` which is caught at compile time or do `request.provide(&self.backtrace)` which will just correctly compare the typeID to a `Backtrace`, which might still be wrong if you were actually trying to return a SpanTrace, like if you do `&self.backtrace` twice and meant to do `&self.span_trace` the second time. It should probably panic if it is provided the same type twice.